### ISSUE TITLE
[AIR 51.1][iOS] `FileStream::openAsync()` inside `Worker` cause hang

### ISSUE THREAD

**Issue by itlancer** [airsdk/Adobe-Runtime-Support#3394]
## Problem Description
`FileStream::openAsync()` call inside `Worker` cause hang for iOS patform using AIR 51.1.

Tested with AIR 51.1.1.1 and AIR 51.1.1.2 with multiple different iOS devices with different OS versions and with different applications.
Same problem in all cases.
There is no such problem using synchronous method to work with `FileStream`.
There is no such problem with AIR 51.0.1.5 and below.
Also there is no such issue with other platforms.
Didn't tested with tvOS.

Related issue: https://github.com/airsdk/Adobe-Runtime-Support/issues/2442

## Steps to Reproduce
Launch application with code below with any iOS device. Application open file asynchronously in `Worker`.

Application example with sources and Scout logs attached.
[air51_ios_worker_bug.zip](https://github.com/user-attachments/files/16494592/air51_ios_worker_bug.zip)

```
package {
	import flash.display.Sprite;
	import flash.system.Worker;
	import flash.system.WorkerDomain;
	import flash.events.Event;
	import flash.filesystem.File;
	import flash.filesystem.FileStream;
	import flash.filesystem.FileMode;
	
	public class AIR51IOSWorkerBug extends Sprite {
		private var worker:Worker;
		private var sprite:Sprite = new Sprite();
		
		public function AIR51IOSWorkerBug() {
			if (Worker.current.isPrimordial) {
				trace("main");
				addEventListener(Event.ADDED_TO_STAGE, init);
			} else {
				trace("worker");
				
				var fileStream:FileStream = new FileStream();
				fileStream.openAsync(File.applicationStorageDirectory.resolvePath("test.txt"), FileMode.APPEND);//This line cause application hang
				//fileStream.open(File.applicationStorageDirectory.resolvePath("test.txt"), FileMode.APPEND);//Use sync mode as a workaround
				fileStream.close();
			}
		}
		
		private function init(e:Event):void {
			removeEventListener(Event.ADDED_TO_STAGE, init);
			
			sprite.graphics.beginFill(0xff0000);
			sprite.graphics.drawRect(0, 0, 100, 100);
			sprite.graphics.endFill();
			sprite.x = 100;
			sprite.y = 100;
			addChild(sprite);
			
			addEventListener(Event.ENTER_FRAME, enterFrame);
			
			worker = WorkerDomain.current.createWorker(this.loaderInfo.bytes, true);
			worker.start();
		}
		
		private function enterFrame(e:Event):void {
			sprite.rotation += 1;
		}
	}
}
```

**Actual Result:**
Application hang. No exceptions, no errors.
Visually application  could works few frames, than hang and nothing happens. Main thread in Scout also hangs at this moment. But Worker thread continue to work.

**Expected Result:**
Application works without hangs.

## Known Workarounds
1) Use synchronous methods to work with files via `FileStream` using `Worker`.
2) Use AIR AIR 51.0.1.5 or below.

**Comment by itlancer** [comment]
@ajwfrost 
With synchronous `FileStream` work the same hang.
`FileStream::writeUTFBytes()` cause hang.
Sample:
```
package {
	import flash.display.Sprite;
	import flash.system.Worker;
	import flash.system.WorkerDomain;
	import flash.events.Event;
	import flash.filesystem.File;
	import flash.filesystem.FileStream;
	import flash.filesystem.FileMode;
	
	public class AIR51IOSWorkerBug extends Sprite {
		private var worker:Worker;
		private var sprite:Sprite = new Sprite();
		
		public function AIR51IOSWorkerBug() {
			if (Worker.current.isPrimordial) {
				trace("main");
				addEventListener(Event.ADDED_TO_STAGE, init);
			} else {
				trace("worker");
				
				var fileStream:FileStream = new FileStream();
				fileStream.open(File.applicationStorageDirectory.resolvePath("test.txt"), FileMode.APPEND);
				fileStream.writeUTFBytes("test");//This line cause application hang
				fileStream.close();
			}
		}
		
		private function init(e:Event):void {
			removeEventListener(Event.ADDED_TO_STAGE, init);
			
			sprite.graphics.beginFill(0xff0000);
			sprite.graphics.drawRect(0, 0, 100, 100);
			sprite.graphics.endFill();
			sprite.x = 100;
			sprite.y = 100;
			addChild(sprite);
			
			addEventListener(Event.ENTER_FRAME, enterFrame);
			
			worker = WorkerDomain.current.createWorker(this.loaderInfo.bytes, true);
			worker.start();
		}
		
		private function enterFrame(e:Event):void {
			sprite.rotation += 1;
		}
	}
}
```

So `FileStream` in `Worker` with iOS completely broken with AIR 51.1 right now.

**Comment by itlancer** [comment]
@ajwfrost 
Issue still exists with latest AIR 51.1.1.3.

**Comment by Hackquarius** [comment]
I can confirm my game hangs on AIR 51.1.+ for no reason on startup. Newest version works good on Desktop and Android

Using AIRSDK 50.2 seemed to worked to package a release for iOS. Apple accepted the .IPA the 1st version of my game is set to release soon. :)

**Comment by ajwfrost** [comment]
@Hackquarius do you think it's this same issue? The other potential start-up issue could be around use of EncryptedLocalStore. We'll be trying to get a fix out for both this and the outstanding ELS things soon...

**Comment by Hackquarius** [comment]
Hi, 

@ajwfrost I'm not sure if it's the same issue, but I tried removing Launchscreen.storyboardc on iOS based on another thread's suggestion. However, it didn't seem to fix the hang-up.

After reviewing my code, I noticed this line at the end of the startup loading process for the main menu of my game:

saveDataObject = SharedObject.getLocal("LoadAllStats");

I usually load all the SharedObject data from memory, and it works fine. Do I need to switch to EncryptedLocalStore on iOS now with the new AIR 51 SDK?

**Comment by Hackquarius** [comment]
@ajwfrost quick update, I have commented this code out

// saveDataObject = SharedObject.getLocal("LoadAllStats");

Now, with iOS on AIR SDK 51, the app no longer hangs, but I still need to load the saved data for my players. :)

**Comment by ajwfrost** [comment]
Okay that's interesting / odd !! We'll check that too, might need a new bug for it if SharedObject.getLocal is causing it to hang too. Might be the same root cause though.. thank you for the details!

**Comment by Hackquarius** [comment]
Hi @ajwfrost , 

Just a simple code snippet with no workers involved. Maybe it's related to the Starling Asset Manager?

Before initializing the Game class from Starling, I create a public var.
public var saveDataObject:SharedObject;

then the saveDataObject is called using SharedObject.getLocal("LoadAllStats36") after loading asset.

There are various data objects handledright after, like checking


			if (saveDataObject.data.HeroLoad_XPTotal == null)
			{ 
				
			}
			else
			{ 
				
				XP_Total = saveDataObject.data.HeroLoad_XPTotal;	
				
			}

**Comment by ajwfrost** [comment]
@Hackquarius  we're not able to reproduce this freeze .. can't see anything in the SharedObject handling that might trigger asynchronous updates, plus with only one thread involved anyway, it shouldn't actually hit anything that would lock it.

So I can't actually see why your case might hang, and I can't see how our fix for the original issue here would impact anything other than Worker instances.

If you're able to create a simple test case that exhibits the problem as a standalone SWF, and send that to us (or attach it here), we can try to reproduce it. But just adding the above code snippets to our constructor is not causing any problems, it seems to work as expected.

thanks

**Comment by ajwfrost** [comment]
Looks good! Might have a play on that next week :-)

**Comment by RossD20Studios** [comment]
Hi @Hackquarius - I believe I, too, have encountered your bug as today was the first time I tried to publish my game, Abalon, to iOS with the latest AIR 51.1. (Previously, worked great in 50.2).

To help shed some light on possible causes, here's what I observed:

1. My game immediately locks up during the "Loading Assets" phase when trying to run the game on anything but a debug connection. If I try to load in debugger to identify the error, the game loads correctly.
2. Based on your notes above, I suspect we have some similarities in our game structure. Specifically, during my loading phase, I am loading a SQLLite database containing our game definitions. This database is loaded by both the Main thread and the backworker thread. Our AI engine runs on the backworker, so it needs to load the definitions stored in the database just like the player in order to understand how to play the game. It appears that this is the point at which the game is failing to load.
3. I am not using an asynchronous connection to load the database, but rather a sync connection as you can see in my code snippet below: <img width="677" alt="Screenshot 2024-09-04 at 12 16 28 PM" src="https://github.com/user-attachments/assets/c3a3e41f-48b5-4315-8b6e-bf05025f04f2">

Hope this helps! I'll continue investigating and provide additional notes here on my findings.

**Comment by Hackquarius** [comment]
Just a quick note about SharedObject.getLocal("LoadAllStats36"). When it was commented out, the debug threw an error, but I was still able to continue, and the loops worked fine. I had been building a standard build for iOS earlier and didn't notice the error because I hadn't tried debugging first.

**Comment by RossD20Studios** [comment]
Thanks for the update @Hackquarius and congrats on your game launch!

@ajwfrost I tested the new AIR SDK today as well and pleased to say it did fix the loading issue I was experiencing. I also played some battles vs. the AI (which uses the backworker) and these seemed to work fine, too. Our AI uses many loops, so I'm not sure how @Hackquarius issue relates, perhaps it is specific to for loops within the loading sequence?

**Comment by ajwfrost** [comment]
@Hackquarius thanks for the reminder about this one.. I have just re-read your description above and am somewhat at a loss!

> I discovered that everything works fine when loading assets, except when it involves assets loaded within a loop. For example:
> `for (y77 = 1; y77 <= Number_Asteroids1_MAX; y77++) {}`
> In my case, the variable Number_Asteroids1_MAX has 500 asteroids. This works perfectly on AIR 51.4 for PC and Android, but iOS seems to hang whenever the loop is executed. I added a trace in the code, and it shows that the loop continues infinitely if Number_Asteroids1_MAX is 2 or more. However, when I set Number_Asteroids1_MAX = 1, the loop skips to the next one, but then it hangs again.

Do you mean that if that `Number_Asteroids1_MAX` is 500, the `y77` variable then continues counting up beyond that limit? This seems a fairly fundamental problem !!
What are the types of those variables?

If you're able to create a (simple as possible) SWF file that demonstrates the issue, we can try to reproduce this and see what's going wrong...

thanks

**Comment by Hackquarius** [comment]
Hi @ajwfrost 


Let me simplify the problem I'm experiencing after further investigation. Any loop with a number greater than 1 encounters an issue when loading from the Starling Asset Manager, resulting in an infinite loop.

For example, if I use a hardcoded loop with a maximum value of 7, the code hangs:

actionscript
Copy code
for (bt = 1; bt <= 7; bt++) { 
    // Code repeats infinitely
}
However, when I set the loop to run only once, it works correctly, and the loading continues:

actionscript
Copy code
for (bt = 1; bt <= 1; bt++) { 
    // Code works and moves to the next lines of code
}
The same issue occurs with this loop:

actionscript
Copy code
for (y77 = 1; y77 <= Number_Asteroids1_MAX; y77++) {}
If Number_Asteroids1_MAX is set to 500, the code hangs, but it works fine when Number_Asteroids1_MAX is set to 1.

I have a lot of loops to load objects, so it's odd that this issue only occurs on iOS; I have no problems on Android or Windows.

I hope this clarifies the problem and helps identify the issue more effectively.

**Comment by Hackquarius** [comment]
Here is a video demonstrating the situation that occurs during debugging.


https://github.com/user-attachments/assets/68fda1a8-1e2b-45db-976a-2c1511865d6b


https://github.com/user-attachments/assets/6d7ac679-aa7e-4a65-a878-7805e5afa00b

**Comment by ajwfrost** [comment]
Hi

We're not able to reproduce this .. and it seems pretty fundamental:
```
for (bt = 1; bt <= 7; bt++) { 
    // Code repeats infinitely
}
```
Can you confirm what data type is being used for `bt`? And what happens if you add a `trace("Val = " + bt);` line in the middle there (or similar, in your longer test apps, if you output the values of the loop variable?)

Either the loop variable is somehow not incrementing properly, or the conditional check isn't working properly.. but both of them seem pretty odd. Is it possible to get a simple SWF file that exhibits the problem?

thanks

**Comment by Hackquarius** [comment]
The bt variable is declared with public var bt:int outside the loop. After testing the trace value of bt and other variables, the trace output gets stuck at 2, indicating that bt is not incrementing as expected.

**Comment by ajwfrost** [comment]
Okay so if it's a public member variable .. is there any chance that it's being updated by something that's being called from outside this loop e.g. when you're adding children or similar, are there any listeners that could be being triggered? My guess is that something may be resetting it to '1' repeatedly, so that it increments to 2 because of the loop update, but then within the loop gets reset back to 1...

Are you able to check the value at different points in the loop - or just omit the bulk of the contents within the loop? So e.g.
```
for (bt = 1; bt <= 7; bt++) { 
    trace("Start of loop - bt = " + bt);
    // calls to other stuff.. minimal perhaps?
    trace("End of loop - bt = " + bt);
}
```

Or if that's meant to be a loop variable, it shouldn't need to be a member at all, so:
```
for (var local_bt : int = 1; local_bt <= 7; local_bt++) { 
    trace("Start of loop - bt = " + local_bt);
    // calls to other stuff..
    trace("End of loop - bt = " + local_bt);
}
```

thanks

**Comment by Hackquarius** [comment]
Hey @ajwfrost , 


The bt variable is always 2 with the trace at start and end of loop.


I tested with 

```
for (var local_bt : int = 1; local_bt <= 7; local_bt++) { 
    trace("Start of loop - bt = " + local_bt);
    // calls to other stuff..
    trace("End of loop - bt = " + local_bt);
}
```

It worked without any issues using the second method without the public variable.

I have no problems with AIR SDK 50.2 or earlier on iOS. I've been working on my game, Cosmic Guardians, for a couple of years now and have tested all SDK versions.

I have many public variables, and they all seem to work with AIR SDK 51, except for the for loops inside the Starling asset manager. It works well on Windows and Android.

**Comment by Hackquarius** [comment]
Hi @ajwfrost ,

I’ve sent the debug SWF and included additional files to help identify the issue.

If you need anything else or require further testing, please let me know.

Have a great day!

**Comment by ajwfrost** [comment]
Thanks

So your loop at the end looks like:
```
     6869   pushstring    	"loaded game2:"  //stringIndex = 3760
     6872   getlex        	y77 //nameIndex = 533
     6875   add           	
     6876   call          	(1)
     6878   pop           	
     6879   findpropstrict	y77 //nameIndex = 533
     6882   getproperty   	y77 //nameIndex = 533
     6885   OP_0x7A       	
     6886   increment     	
     6887   findpropstrict	y77 //nameIndex = 533
     6890   swap          	
     6891   setproperty   	y77 //nameIndex = 533
     6894   debugline     	12256
     
     L1: 
     6897   getlex        	y77 //nameIndex = 533
     6900   getlex        	Number_Asteroids1_MAX //nameIndex = 1906
     6903   ifle          	L2
```

Instructions 6869-6878 get the value of y77, add (concat) it to "loaded game 2:", call trace and discard the void result.
6879-6891 _should_ find 'y77' and get its value, increment it, and then set it again.
6897-6903 then get the y77 value, and the loop max, and if y77 is less than Number_Asteroids1_MAX, then we loop back to the top (L2 label).

The obvious concern is OP_0x7A which SWFInvestigator doesn't know about .. looking in our code, this is `unplus` 
[see avmplus interpreter](https://github.com/adobe/avmplus/blob/858d034a3bd3a54d9b70909386435cf4aec81d21/core/Interpreter.cpp#L459)

And it's not immediately obvious from the name, but that instruction _should_  just make sure a value is a numeric type.. so e.g. null turns into zero, true turns into 1, a string is parsed to a number, etc.
`// OP_unplus is an instruction (unary plus; "coerce to numeric")`

However, if that was returning 1, then you would probably end up with this functionality i.e. instruction 6885 sets it to one, instruction 6886 increments this to two, and then you compare it and start the loop again.

I think we'll need to build some extra profiling into the compile-abc-64 tool and run your SWF through this, to see what machine code is being created from this ABC code...

thanks

**Comment by ajwfrost** [comment]
Interesting: if I build some similar code, I get:
```
       43   getlex        	loopvar //nameIndex = 1
       45   convert_d     	
       46   increment     	
       47   findpropstrict	loopvar //nameIndex = 1
       49   swap          	
       50   setproperty   	loopvar //nameIndex = 1
```
i.e. `convert_d` rather than `unplus`.

Ah - but as soon as we add a `float` variable elsewhere into the code, we see this change to `unplus` (I think the compiler is checking whether we need float support...)

Okay and this is when I start wonder how it ever worked. When I create an IPA from this, it throws an error saying it can't create the loop variable property. Adding some debug into the ABC-to-machine-code conversion, and we are missing some opcodes..! Which means that this definitely seems to be the problem..

Looks like you're using Windows there? We can try to get an updated version of compile-abc-64.exe to try out a fix here...

thanks

**Comment by Hackquarius** [comment]
Yes, I always use Windows, except when creating certificates for iOS and exporting the .IPA to APPTransporter on macOS using MACinCloud. Thank you for identifying the issue, and let me know if I can assist further.

**Comment by ajwfrost** [comment]
@Hackquarius  apologies for the delay; below is a zipped-up version of the "compile-abc-64.exe" tool that should go into your SDK under the "lib\aot\bin\compile-abc\" folder. If you could try using this and see if it helps; for us, this has corrected the machine code generation so that our test case is now working properly..

thanks
[compile-abc-64.zip](https://github.com/user-attachments/files/17261758/compile-abc-64.zip)

**Comment by Hackquarius** [comment]
Hi @ajwfrost ,

After testing the latest AIR SDK with the compile-abc-64 file, I can confirm that the game loads well and the issue is resolved! 

Thank you very much to you and the Harman team for resolving this problem on iOS!

**Comment by ajwfrost** [comment]
Great, thanks for the quick confirmation! We'll get this into our forthcoming release..

**Comment by itlancer** [comment]
Original issue also fixed with AIR 51.1.1.5.
Thanks!
